src/libexpr/eval.cc

```diff
@@ -2017,7 +2018,11 @@ BackedStringView EvalState::coerceToString(const Pos & pos, Value & v, PathSet &
         if (v.type() == nBool && v.boolean) return "1";
         if (v.type() == nBool && !v.boolean) return "";
         if (v.type() == nInt) return std::to_string(v.integer);
-        if (v.type() == nFloat) return std::to_string(v.fpoint);
+        if (v.type() == nFloat) {
+            std::stringstream out;
+            out << std::setprecision(15) << v.fpoint;
```

Comment:
15 digits does not suffice to distinguish some doubles, such as `1` and `1.0000000000000002`. This should be `std::numeric_limits<double>::max_digits10 == 17`.

Reply:
that is true. in the interest of keeping at least somewhat close to the original implementation we chose to print floats with fewer digits than the binary format supports to avoid least significant digit artifacts like `1.1` printing as `1.1000000000000001`. it seemed like a good compromise between accuracy and potential impact of system-dependent rounding artifacts on the output.

Reply:
If we’re going to risk changing evaluation results for this, maybe we should pull in the [Ryū](https://github.com/ulfjack/ryu) library and do it right, so we don’t have to change them again later?

In the future when we can rely on GCC ≥ 11, we’ll be able to use `std::to_chars`, which uses a Ryū implementation built into libstdc++.

Reply:
that would indeed be much better. going to mark this as draft for now until we can give this the attention such a change deserves.